File: samples/java8-sample-contract/src/main/java/my/sample/java8/contract/rockps/domain/game/Game.java
package my.sample.java8.contract.rockps.domain.game;

import my.sample.java8.contract.rockps.util.Util;

import java.time.OffsetDateTime;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Game {
    private Map<String, Player> players;
    private Player winner;
    private GameStatus status = GameStatus.ACTIVE;

    public Game() {
    }

    public Map<String, Player> getPlayers() {
        return players;
    }

    public Player getWinner() {
        return winner;
    }

    public Game(List<Player> players) {
        this.players = players.stream().collect(Collectors.toMap(Player::getAddress, Function.identity()));
    }

    public void reveal(String address, String salt) {
        Player player = players.get(address);
        if (player == null) {
            throw new IllegalArgumentException("Address " + address + " isn't among players of the game");
        }
        Map<String, AnswerType> hashedAnswers = getHashedAnswers(salt);
        AnswerType answer = hashedAnswers.get(player.getHashedAnswer());
        if (answer == null) {
            throw new IllegalArgumentException("Not found matching answer for salt");
        }
        player.setAnswer(answer);
        if (allPlayersAnswered()) {
            finish();
        }
    }

    private void finish() {
        status = GameStatus.FINISHED;
        if (allPlayersHaveTheSameAnswers()) {
            return;
        }
        this.winner = players.values().stream().max(
                (player1, player2) -> new AnswerComparator().compare(player1.getAnswer(), player2.getAnswer())
        ).orElseThrow(() -> new IllegalStateException("Winner could not be determined"));
    }

    private boolean allPlayersHaveTheSameAnswers() {
        Set<AnswerType> differentAnswers = players.values().stream().map(Player::getAnswer).collect(Collectors.toSet());
        return differentAnswers.size() == 1;
    }

    private boolean allPlayersAnswered() {
        return players.values().stream().allMatch(player -> player.getAnswer() != null);
    }

    public Map<String, AnswerType> getHashedAnswers(String salt) {
        return Arrays.stream(AnswerType.values()).collect(
                Collectors.toMap(answerType -> Util.hash(answerType + "_" + salt), Function.identity())
        );
    }

    public GameStatus getStatus() {
        return status;
    }

    static class AnswerComparator implements Comparator<AnswerType> {

        @Override
        public int compare(AnswerType o1, AnswerType o2) {
            if ((o1 == AnswerType.SCISSORS && o2 == AnswerType.ROCK) || (o1 == AnswerType.ROCK && o2 == AnswerType.SCISSORS)) {
                return -o1.compareTo(o2);
            }
            return o1.compareTo(o2);
        }
    }
}

File: samples/kotlin-sample/kotlin-contract/src/main/kotlin/my/sample/kotlin/contract/rockps/game/Game.kt
package my.sample.kotlin.contract.rockps.game

import my.sample.kotlin.contract.rockps.util.hash

data class Game(
    val players: Map<String, Player>,
    var winner: Player? = null,
    var status: GameStatus = GameStatus.ACTIVE,
) {

    fun reveal(address: String, salt: String) {
        val player = requireNotNull(players[address]) {
            "Address $address isn't among players of the game"
        }
        val hashedAnswers = getHashedAnswers(salt)
        val answer = requireNotNull(hashedAnswers[player.hashedAnswer]) {
            "Not found matching answer for salt"
        }
        player.answer = answer
        if (allPlayersAnswered()) {
            finish()
        }
    }

    private fun finish() {
        status = GameStatus.FINISHED
        if (allPlayersHaveTheSameAnswers()) {
            return
        }
        winner = players.values.maxWithOrNull { a: Player, b: Player ->
            AnswerComparator().compare(a.answer!!, b.answer!!)
        } ?: throw IllegalStateException(
            "Winner could not be determined"
        )
    }

    private fun allPlayersHaveTheSameAnswers(): Boolean {
        val differentAnswers = players.values.map { obj: Player -> obj.answer }.toSet()
        return differentAnswers.size == 1
    }

    private fun allPlayersAnswered() = players.values.all { player: Player -> player.answer != null }

    private fun getHashedAnswers(salt: String) =
        AnswerType.values().associateBy { hash(it.toString() + "_" + salt) }

    internal class AnswerComparator : Comparator<AnswerType> {
        override fun compare(o1: AnswerType, o2: AnswerType): Int {
            return if (o1 == AnswerType.SCISSORS && o2 == AnswerType.ROCK || o1 == AnswerType.ROCK && o2 == AnswerType.SCISSORS) {
                -o1.compareTo(o2)
            } else o1.compareTo(o2)
        }
    }
}